### ISSUE TITLE
Convert to TypeScript and add public types

### ISSUE THREAD

**Issue by crosscompile** [palmerhq/cypress-image-snapshot#75]
There have been a few issues related to TypeScript types (#13, #72), updating the source code to TypeScript would make future changes easier and give consumers better type support.

**Comment by jordandlaman** [comment]
So admittedly I am a TS amateur.  I was just playing around with things and I did get it to config a bit easier if I added index.d.ts directly to the node_modules/cypress-image-snapshot/ with
``` javascript
import 'cypress';

declare global {
  namespace Cypress {
    interface Chainable {
      matchImageSnapshot: (options?: any) => void;
    }
  }
}
```
That removed the need to include a custom types file and include that in my tsconfig.json file for the cypress tests and or using 
``` javascript
/// <reference types="../types/cypress-image-snapshot/" />
```
in the spec file to reference the type file manually.  I then can use the import statement in my support/index.ts like so
``` javascript
import 'cypress-image-snapshot/command'
```
Obviously hacking a node_modules folder is a no no but I thought that might be a way to go.  I have seen a few other plugins use this sort of pattern and thought I would give it a shot.

**Comment by crosscompile** [comment]
Yep that makes sense!  So if we add an `index.d.ts` to the root of cypress-image-snapshot and publish to NPM again, everyone will receive types without having to touch their `node_modules` directly.

I think what would be even better is if we rewrite to use TypeScript and then generate `index.d.ts` from our source, similar to how [Formik](https://github.com/jaredpalmer/formik) does.

But creating types manually is definitely useful if we want to start there.

**Comment by makeupsomething** [comment]
Im pretty new to typescript but would like to have a go at this issue if it is ok?

**Comment by crosscompile** [comment]
@makeupsomething for sure give it a shot!  I'm happy to review any PRs.

**Comment by ell10t** [comment]
```ts
import "cypress";
import { MatchImageSnapshotOptions } from "jest-image-snapshot";

declare global {
    namespace Cypress {
        interface Chainable {
            matchImageSnapshot: (
                options?:
                    | string
                    | (Partial<
                          MatchImageSnapshotOptions & Loggable & Timeoutable & ScreenshotOptions
                      >),
            ) => void;
        }
    }
}
```

Just based on what I understood from the readme, just a good stop gap for now.

**Comment by anujpunia** [comment]
> So admittedly I am a TS amateur. I was just playing around with things and I did get it to config a bit easier if I added index.d.ts directly to the node_modules/cypress-image-snapshot/ with
> 
> ```js
> import 'cypress';
> 
> declare global {
>   namespace Cypress {
>     interface Chainable {
>       matchImageSnapshot: (options?: any) => void;
>     }
>   }
> }
> ```
> 
> Obviously hacking a node_modules folder is a no no but I thought that might be a way to go though im not sure how much different it is from having a custom types folder somewhere else. I have seen a few other plugins use this sort of pattern and thought I would give it a shot.

Adding `index.d.ts` as you said still gives me error in the .ts spec file- 
`Property 'matchImageSnapshot' does not exist on type 'Chainable<undefined>'.`

**Comment by Lukaszli** [comment]
Any solution? How to use this plugin with typescript?

**Comment by edupeeth** [comment]
I shifted to Percy. They have a free version as well and much smoother. Had no luck here.

**Comment by Keysox** [comment]
I went ahead and added typings for folks to use: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41222 

Now, all you need to do is `yarn add @types/cypress-image-snapshot -D` for this plugin to work with TypeScript!

**Comment by Svish** [comment]
@Keysox Think the options are missing `Cypress.ScreenshotOptions`

**Comment by Keysox** [comment]
@Svish -- options have been added if you upgrade to `3.1.1` https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41897/files#diff-603108fb86a139bc0934ce518b4267eeR10

**Comment by StallionV** [comment]
I still cannot get this to work with typescript.  I followed instructions on the [main page](https://github.com/palmerhq/cypress-image-snapshot) 
But get an error
`Property 'matchImageSnapshot' does not exist on type 'cy & EventEmitter'.ts(2339)`
I tried adding types` @types/cypress-image-snapshot` as suggested in one of the other thread but still cannot get it to work.
I changed the test file from .ts to .js to get it to run but then get the following error
```
cy.task('Matching image snapshot') failed with the following error:

The task 'Matching image snapshot' was not handled in the plugins file. The following tasks are registered: resetCoverage, combineCoverage, coverageReport

```

Any pointers?

**Comment by ar5had** [comment]
@Keysox For some reason, even after installing the types, matchImageSnapshot property seems to be missing :(
